### PR TITLE
Refactor the lexer to be less implicit

### PR DESCRIPTION
**Pull request by NoraCodes**
Just some ideas regarding factorization. This makes it a lot easier to e.g. have the lexer emit its own state on weird conditions. I'd also like to move the match arms into methods on `Lexer` but have to study for my algos exam.

### CODE REVIEW

**Comment by CosineP** [comment]
woah, thanks a lot! i didn't expect to get any PRs (at least this early on). i'll take a closer look when i get back in a few minutes but im wondering if you could expand on the logic here:

1. what does it mean you can emit its own state on weird conditions? Can't you emit the same amount of state as u can now by printing rchars.last()?
2. why putting everything in an impl?
3. check it out, this is closer to how i was doing it before using rchars like a stack. https://github.com/CosineGaming/sfg/blob/5803e916dd99a29b22abc56066d91eb4ec1b1051/src/lib.rs#L33 i ultimately went *away* from this design because it was frustrating to work in two places at once, so im wary. You can still convince me probably though. i can tell that still using rchars makes it cleaner now, but still

**Comment by CosineP** [comment]
also the project uses tabs

**Comment by NoraCodes** [comment]
> also the project uses tabs

RIP me and my Vimconfig, sorry about that!

I'll fix that and explain more when I get back from the study session

**Comment by CosineP** [comment]
hot tip: https://github.com/tpope/vim-sleuth is an excellent vimscript that'll make you never have to think about spaces or tabs again. i miss it in kakoune

File: rsfg/src/lexer.rs [hunk 1]
```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
...
 }
+
 // A-Z or 0-9
 fn is_id(c: char) -> bool {
-	is_id_1st(c) || (c >= '0' && c <= '9')
+    is_id_1st(c) || (c >= '0' && c <= '9')
+}
+
+enum NextSymbolType {
+    None,
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: name clash with `Option::None`, maybe `EOF`?

File: rsfg/src/lexer.rs [hunk 2]
```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
...
+    SymbolOrId(char),
+    Space(char),
+    Digit(char),
+    Slash,
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: DivideOrComment based on SymbolOrId?

File: rsfg/src/lexer.rs [hunk 3]
```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
...
+    }
+
+    fn next_symbol_type(&mut self) -> NextSymbolType {
+        use NextSymbolType::*;
+
+        if self.is_eof() {
+            return None;
+        }
+
+        let c = self.rchars.pop().expect("not supposed to be at EOF");
```
Comment (CosineP) [inline rsfg/src/lexer.rs]:
```rust
let c = match self.rchars.pop() {
    Some(c) => c,
    None => return None, // or EOF
};
```
agrees with how i do everything else, and then u can drop the `is_eof` function

File: rsfg/src/lexer.rs [hunk 4]
```diff
@@ -6,175 +6,235 @@ static SPACE: &str = " \t";
...
+                        }
+                        continue;
+                    }
+                    _ => {
+                        // Division
+                        panic!("division operator not yet supported");
+                    }
+                }
+            }
+            NextSymbolType::Newline => Newline,
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: i guess this here is the real payoff, not having to rchars.pop() every time, so nice

**Comment by CosineP** [comment]
also consider if this makes things cleaner without separating the first character into a separate function: https://github.com/CosineGaming/sfg/commit/e40e5f15e73b282d0ce1433fcb1c27388b571e90

File: rsfg/src/lexer.rs (hunk 1 above)
Comment (NoraCodes) [reply rsfg/src/lexer.rs]: Good call, done.

File: rsfg/src/lexer.rs (hunk 3 above)
Comment (NoraCodes) [reply rsfg/src/lexer.rs]: Yes, definitely nicer. Done.

**Comment by NoraCodes** [comment]
In terms of efficiency, it's tempting to either use a ring buffer (`VecDeque`) or even remove that stage of copying by just having an index into the &'str that's passed in.

On the other hand, it works as is. Would you mind if I were to implement that?

**Comment by CosineP** [comment]
well you know what they say, premature optimization is the root of all evil. If it ends up being a bottleneck, id love to have it cause idk how to do that stuff. But if not i'd rather save the complexity, especially since speed is only a secondary goal with this project

**Comment by NoraCodes** [comment]
In that case I think this is basically done, I'll remove the [WIP].

File: rsfg/src/lexer.rs [hunk 5]
```diff
@@ -2,146 +2,187 @@
...
-fn is_id_1st(c: char) -> bool {
-	c >= 'A' && c <= 'z'
-}
 // A-Z or 0-9
 fn is_id(c: char) -> bool {
-	is_id_1st(c) || (c >= '0' && c <= '9')
+	match c {
+		'0'..='9' | 'A'..='z' => true,
+		_ => false
+	}
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: nah, this matches on brackets and underscores and such. https://www.cs.cmu.edu/~pattis/15-1XX/common/handouts/ascii.html

File: rsfg/src/lexer.rs [hunk 6]
```diff
@@ -2,146 +2,187 @@
...
+		}
+	}
+
+	fn next_symbol_type(&mut self) -> NextTokenType {
+		use NextTokenType::*;
+
+		match self.rchars.pop() {
+			None => EOF,
+			Some(c) => match c {
+				'A'..='z' => SymbolOrId(c),
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: still matches too much

File: rsfg/src/lexer.rs [hunk 7]
```diff
@@ -2,146 +2,187 @@
...
-					Some(x) => *x,
-					None => break, // End of ID is fine
-				};
-				if is_id(x) {
-					text.push(x);
-				} else {
-					break;
-				}
-				rchars.pop();
+		println!("{:?}", lexer);
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: oh okay, i see this is a demonstration of how bundling things into a struct makes debugging easier. i like it! although it's only 4 variables.... either way this should come out when we're done

File: rsfg/src/lexer.rs [hunk 8]
```diff
@@ -2,146 +2,187 @@
...
+		let token = match lexer.next_symbol_type() {
+			NextTokenType::EOF => {
+				// This is the end of the file, which is OK, as we are not in the middle
+				// of matching a token
+				break;
```
Comment (CosineP) [inline rsfg/src/lexer.rs]: nice, this is clearer

File: rsfg/src/lexer.rs (hunk 7 above)
Comment (NoraCodes) [reply rsfg/src/lexer.rs]: Yeah, definitely.

File: rsfg/src/lexer.rs (hunk 5 above)
Comment (NoraCodes) [reply rsfg/src/lexer.rs]: Ope, that's a bad mistake. Should be `'0'..='9' | 'A'..='Z' | 'a'..='z'`, sorry.

**Comment by CosineP** [comment]
i hope you don't mind, i needed to make some lexer changes so to avoid merge conflicts i just went ahead and merged and will fix the issues myself

**Comment by NoraCodes** [comment]
Totally, sorry for taking forever on this. It's midterms and I'm swamped :(

**Comment by CosineP** [comment]
oh no problem are you kidding, you volunteered your help which i never expected xD

**Pull request merged**